File: baml_language/sdk_tests/crates/swift/type_shapes/customizable/roundtrip_tests/TestLists.swift
// Roundtrip coverage for `Baml.lists` — port of python_pydantic2
// `roundtrip_tests/test_lists.py`.
//
import XCTest
import Baml
import BamlBridge

final class TestLists: XCTestCase {
    func test_lists_round_trip_union_list() throws {
        // `(int | string)[]` → `[BamlUnion2<Int, String>]`.
        let xs: [BamlUnion2<Int, String>] = [.t0(1), .t1("two"), .t0(3)]
        XCTAssertEqual(try Baml.lists.round_trip_union_list(xs: xs), xs)
    }

    func test_lists_round_trip_list_container() throws {
        let c = Baml.lists.ListContainer(
            ints: [1, 2],
            optional_strings: [nil, "z"],
            union_list: [.t0(1), .t1("x")]
        )
        XCTAssertEqual(try Baml.lists.round_trip_list_container(c: c), c)
    }

    func test_lists_round_trip_ints() throws {
        XCTAssertEqual(try Baml.lists.round_trip_ints(xs: [1, 2, 3]), [1, 2, 3])
    }

    func test_lists_round_trip_empty_list() throws {
        // Mirrors Python's Bug A regression: an empty list must encode
        // as a present-but-empty list_value, not an unset oneof (which
        // the engine reads as null).
        XCTAssertEqual(try Baml.lists.round_trip_ints(xs: []), [])
    }

    func test_lists_round_trip_optional_strings() throws {
        XCTAssertEqual(
            try Baml.lists.round_trip_optional_strings(xs: ["a", nil, "b"]),
            ["a", nil, "b"]
        )
    }
}

File: baml_language/sdk_tests/crates/java/type_shapes/customizable/roundtrip_tests/TestLists.java
// Roundtrip coverage for `baml_sdk.lists` — list Ty variants.
//
// Port of python_pydantic2/type_shapes/customizable/roundtrip_tests/
// test_lists.py — same test names, cases, inputs, assertions.
//
// java-port note: for the union-shaped elements below, see TestUnions.java
// for the generic-family shape (`(int | string)` -> `Union2<Long, String>`,
// int = Arm0, string = Arm1) this port assumes.
package roundtrip_tests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import baml_bridge.Union2;
import baml_sdk.lists.Fns;
import baml_sdk.lists.ListContainer;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.Test;

class TestLists {

    @Test
    void test_lists_round_trip_ints() {
        assertEquals(List.of(1L, 2L, 3L), Fns.round_trip_ints(List.of(1L, 2L, 3L)));
    }

    @Test
    void test_lists_round_trip_empty_list() {
        // Regression for Bug A (35b), fixed by `SetInParent()` on the
        // `list_value` oneof arm in proto.py: an empty list used to encode
        // as an unset oneof, which the engine read as null and returned as
        // `None`.
        assertEquals(List.of(), Fns.round_trip_ints(List.of()));
    }

    @Test
    void test_lists_round_trip_optional_strings() {
        // java-port note: `List.of` rejects null elements; `Arrays.asList`
        // permits them and is used wherever a list literal carries a
        // `null` element.
        List<String> xs = Arrays.asList("a", null, "b");
        assertEquals(xs, Fns.round_trip_optional_strings(xs));
    }

    @Test
    void test_lists_round_trip_union_list() {
        List<Union2<Long, String>> xs =
                List.of(
                        new Union2.Arm0<Long, String>(1L),
                        new Union2.Arm1<Long, String>("two"),
                        new Union2.Arm0<Long, String>(3L));
        assertEquals(xs, Fns.round_trip_union_list(xs));
    }

    @Test
    void test_lists_round_trip_list_container() {
        ListContainer c =
                new ListContainer(
                        List.of(1L, 2L),
                        Arrays.asList(null, "z"),
                        List.of(
                                new Union2.Arm0<Long, String>(1L),
                                new Union2.Arm1<Long, String>("x")));
        assertEquals(c, Fns.round_trip_list_container(c));
    }
}
